// checkout-sdk-node/dist/api/reconciliation/reconciliation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
/**
 * Class dealing with the /reporting endpoint
 *
 * @export
 * @class Reconciliation
 */
class Reconciliation {
  constructor(config) {
    this.config = config;
  }

  /**
   * Returns a JSON report containing all payments within your specified parameters
   *
   * @memberof Reconciliation
   * @param {Object} body Reconciliation request body.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getPayments(body) {
    var _this = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this.config.host}/reporting/payments`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this.config.httpClient, url, _this.config, _this.config.sk);
        const res = yield response.json;

        // In case there is a "next" page, inject it in the response body
        if (res._links && res._links.next) {
          const nextLink = res._links.next.href;
          return yield _objectSpread(_objectSpread({}, res), {}, {
            page: nextLink.match(/after=([^&]*)/)[1]
          });
        }
        return yield res;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns a JSON payment report containing all of the data related to a specific payment,
   * based on the payment's identifier.
   *
   * @memberof Reconciliation
   * @param {string} paymentId Payment id.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getPayment(paymentId) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this2.config.httpClient, `${_this2.config.host}/reporting/payments/${paymentId}`, _this2.config, _this2.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns a JSON report containing all payments within your specified parameters
   *
   * @memberof Reconciliation
   * @param {Object} body Reconciliation request body.
   * @return {Promise<Buffer>} A promise to the request reconciliation response.
   */
  getPaymentsCsv(body) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this3.config.host}/reporting/payments/download`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this3.config.httpClient, url, _objectSpread(_objectSpread({}, _this3.config), {}, {
          csv: true
        }), _this3.config.sk);
        return yield response.csv;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns a JSON report containing all statements within your specified parameters.
   * Please note that the timezone for the request will be UTC.
   *
   * @memberof Reconciliation
   * @param {Object} body Reconciliation request body.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getStatements(body) {
    var _this4 = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this4.config.host}/reporting/statements`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this4.config.httpClient, url, _this4.config, _this4.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Downloads a CSV statement report containing all of the data related to a specific
   * statement, based on the statement's identifier.
   *
   * @memberof Reconciliation
   * @param {string} statementId Statement id.
   * @return {Promise<Buffer>} A promise to the request reconciliation response.
   */
  getStatementCsv(statementId) {
    var _this5 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this5.config.httpClient, `${_this5.config.host}/reporting/statements/${statementId}/payments/download`, _objectSpread(_objectSpread({}, _this5.config), {}, {
          csv: true
        }), _this5.config.sk);
        return yield response.csv;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns all associated payment actions that impact your balance within the parameters you specify
   *
   * @memberof Reconciliation
   * @param {Object} body Reconciliation request body.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getPaymentsActions(body) {
    var _this6 = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this6.config.host}/reporting/actions`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this6.config.httpClient, url, _this6.config, _this6.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns the reconciliation data of the payment action
   *
   * @memberof Reconciliation
   * @param {string} actionsId Action id.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getPaymentsAction(actionsId) {
    var _this7 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this7.config.httpClient, `${_this7.config.host}/reporting/payments/actions/${actionsId}`, _this7.config, _this7.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns a CSV report containing all payments within your specified parameters
   *
   * @memberof Reconciliation
   * @param {Object} body Reconciliation request body.
   * @return {Promise<Buffer>} A promise to the request reconciliation response.
   */
  getPaymentsActionsCsv(body) {
    var _this8 = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this8.config.host}/reporting/actions/download`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this8.config.httpClient, url, _objectSpread(_objectSpread({}, _this8.config), {}, {
          csv: true
        }), _this8.config.sk);
        return yield response.csv;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Returns the reconciliation data of a payment action
   *
   * @memberof Reconciliation
   * @param {string} actionId Action id.
   * @return {Promise<Object>} A promise to the request reconciliation response.
   */
  getAction(actionId) {
    var _this9 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this9.config.httpClient, `${_this9.config.host}/reporting/actions/${actionId}`, _this9.config, _this9.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }
}
exports.default = Reconciliation;
