// checkout-sdk-node/dist/api/apm-specific/sepa.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
/**
 * Class dealing with the /sepa and /ppro/sepa endpoint
 *
 * @export
 * @class Sepa
 */
class Sepa {
  constructor(config) {
    this.config = config;
  }

  /**
   * Get mandate
   *
   * @param {string} id Source id
   * @return {Promise<Object>} A promise to the Sepa response.
   */
  getMandate(id) {
    var _this = this;
    return _asyncToGenerator(function* () {
      const url = _this.config.host.includes('sandbox') ? `${_this.config.host}/sepa-external/mandates/${id}` : `${_this.config.host}/sepa/mandates/${id}`;
      try {
        const response = yield (0, _http.get)(_this.config.httpClient, url, _this.config, _this.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Cancel mandate
   *
   * @param {string} id Source id
   * @return {Promise<Object>} A promise to the Sepa response.
   */
  cancelMandate(id) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      const url = _this2.config.host.includes('sandbox') ? `${_this2.config.host}/sepa-external/mandates/${id}/cancel` : `${_this2.config.host}/sepa/mandates/${id}/cancel`;
      try {
        const response = yield (0, _http.post)(_this2.config.httpClient, url, _this2.config, _this2.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Get mandate via PPRO
   *
   * @param {string} id Source id
   * @return {Promise<Object>} A promise to the Sepa response.
   */
  getPPROMandate(id) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      const url = _this3.config.host.includes('sandbox') ? `${_this3.config.host}/ppro/sepa-external/mandates/${id}` : `${_this3.config.host}/ppro/sepa/mandates/${id}`;
      try {
        const response = yield (0, _http.get)(_this3.config.httpClient, url, _this3.config, _this3.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Cancel mandate via PPRO
   *
   * @param {string} id Source id
   * @return {Promise<Object>} A promise to the Sepa response.
   */
  cancelPPROMandate(id) {
    var _this4 = this;
    return _asyncToGenerator(function* () {
      const url = _this4.config.host.includes('sandbox') ? `${_this4.config.host}/ppro/sepa-external/mandates/${id}/cancel` : `${_this4.config.host}/ppro/sepa/mandates/${id}/cancel`;
      try {
        const response = yield (0, _http.post)(_this4.config.httpClient, url, _this4.config, _this4.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }
}
exports.default = Sepa;
