// checkout-sdk-node/dist/api/events/events.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
/**
 * Class dealing with the /events endpoint
 *
 * @export
 * @class Events
 */
class Events {
  constructor(config) {
    this.config = config;
  }

  /**
   * Retrieve a list of event types grouped by their respective version that you can
   * configure on your webhooks.
   *
   * @memberof Events
   * @param {string} version Events Version.
   * @return {Promise<Object>} A promise to the request events response.
   */
  retrieveEventTypes(version) {
    var _this = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this.config.host}/event-types`;
        if (version) {
          url += `?version=${version}`;
        }
        const response = yield (0, _http.get)(_this.config.httpClient, url, _this.config, _this.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Retrieves events ordered by the event date in descending order (latest first).
   * Results can be paged by specifying the skip and limit query parameters.
   *
   * @memberof Events
   * @param {Object} body Events request body.
   * @return {Promise<Object>} A promise to the request events response.
   */
  retrieveEvents(body) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this2.config.host}/events`;
        if (body) {
          const queryString = Object.keys(body).map(key => `${key}=${body[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this2.config.httpClient, url, _this2.config, _this2.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Retrieves the event with the specified identifier string. The event data includes the full event
   * details, the schema of which will vary based on the type.
   *
   * @memberof Events
   * @param {string} eventId Event id.
   * @return {Promise<Object>} A promise to the request event response.
   */
  retrieveEvent(eventId) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this3.config.httpClient, `${_this3.config.host}/events/${eventId}`, _this3.config, _this3.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Retrieves the attempts for a specific event notification
   *
   * @memberof Events
   * @param {Object} body Event request body.
   * @return {Promise<Object>} A promise to the request event notifications response.
   */
  retrieveEventNotification(body) {
    var _this4 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this4.config.httpClient, `${_this4.config.host}/events/${body.eventId}/notifications/${body.notificationId}`, _this4.config, _this4.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Retries a specific webhook notification for the given event
   *
   * @memberof Events
   * @param {Object} body Event request body.
   * @return {Promise<Object>} A promise to the retry event response.
   */
  retry(body) {
    var _this5 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.post)(_this5.config.httpClient, `${_this5.config.host}/events/${body.eventId}/webhooks/${body.webhookId}/retry`, _this5.config, _this5.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Retries all webhook notifications configured for the specified event
   *
   * @memberof Events
   * @param {string} eventId Event id.
   * @return {Promise<Object>} A promise to the retry events response.
   */
  retryAll(eventId) {
    var _this6 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.post)(_this6.config.httpClient, `${_this6.config.host}/events/${eventId}/webhooks/retry`, _this6.config, _this6.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }
}
exports.default = Events;
