// checkout-sdk-node/dist/api/access/access.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
/**
 * Class dealing with the access api
 *
 * @export
 * @class Access
 */
class Access {
  constructor(config) {
    this.config = config;
  }

  /**
   * Request an access token
   *
   * @param {Object} body Access object body.
   * @return {Promise<Object>} A promise to the Access response.
   */
  request(body) {
    var _this = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.createAccessToken)(_this.config, _this.config.httpClient, body);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Update the access details in the config.
   *
   * @param {Object} body Access response body.
   * @return {void} 
   */
  updateAccessToken(body) {
    this.config.access = {
      token: body.access_token,
      type: body.token_type,
      scope: body.scope,
      expires: new Date(new Date().getTime() + body.expires_in)
    };
  }
}
exports.default = Access;
