// checkout-sdk-node/dist/api/reports/reports.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
/**
 * Class dealing with the /reports api endpoint
 *
 * @export
 * @class Reports
 */
class Reports {
  constructor(config) {
    this.config = config;
  }

  /**
   * Returns the list of reports and their details.
   *
   * @memberof Reports
   * @param {Object} parameters Parameters.
   * @return {Promise<Object>} A promise to the Reports response.
   */
  getAllReports(parameters) {
    var _this = this;
    return _asyncToGenerator(function* () {
      try {
        let url = `${_this.config.host}/reports`;
        if (parameters) {
          const queryString = Object.keys(parameters).map(key => `${key}=${parameters[key]}`).join('&');
          url += `?${queryString}`;
        }
        const response = yield (0, _http.get)(_this.config.httpClient, url, _this.config, _this.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Use this endpoint to retrieve a specific report using its ID.
   *
   * @memberof Reports
   * @param {string} id The ID of the report to retrieve.
   * @return {Promise<Object>} A promise to the Reports response.
   */
  getReportDetails(id) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this2.config.httpClient, `${_this2.config.host}/reports/${id}`, _this2.config, _this2.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Use this endpoint to retrieve a specific file from a given report using their respective IDs.
   *
   * @memberof Reports
   * @param {string} id The ID of the report that the file belongs to.
   * @param {Object} fileId The ID of the file to retrieve.
   * @return {Promise<Object>} A promise to the Reports response.
   */
  getReportFile(id, fileId) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.get)(_this3.config.httpClient, `${_this3.config.host}/reports/${id}/files/${fileId}`, _objectSpread(_objectSpread({}, _this3.config), {}, {
          csv: true
        }), _this3.config.sk);
        return yield response.csv;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }
}
exports.default = Reports;
