// checkout-sdk-node/dist/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TRANSFERS_SANDBOX_URL = exports.TRANSFERS_LIVE_URL = exports.SANDBOX_BASE_URL = exports.SANDBOX_ACCESS_URL = exports.REQUEST_ID_HEADER = exports.PLATFORMS_FILES_SANDBOX_URL = exports.PLATFORMS_FILES_LIVE_URL = exports.PAYMENT_TYPES = exports.NAS_SANDBOX_SECRET_KEY_REGEX = exports.NAS_SANDBOX_PUBLIC_KEY_REGEX = exports.NAS_LIVE_SECRET_KEY_REGEX = exports.NAS_LIVE_PUBLIC_KEY_REGEX = exports.MBC_LIVE_SECRET_KEY_REGEX = exports.LIVE_BASE_URL = exports.LIVE_ACCESS_URL = exports.DEFAULT_TIMEOUT = exports.CURRENCIES = exports.BALANCES_SANDBOX_URL = exports.BALANCES_LIVE_URL = exports.API_VERSION_HEADER = void 0;
const SANDBOX_BASE_URL = exports.SANDBOX_BASE_URL = 'https://api.sandbox.checkout.com';
const LIVE_BASE_URL = exports.LIVE_BASE_URL = 'https://api.checkout.com';
const SANDBOX_ACCESS_URL = exports.SANDBOX_ACCESS_URL = 'https://access.sandbox.checkout.com/connect/token';
const LIVE_ACCESS_URL = exports.LIVE_ACCESS_URL = 'https://access.checkout.com/connect/token';
const PLATFORMS_FILES_LIVE_URL = exports.PLATFORMS_FILES_LIVE_URL = 'https://files.checkout.com/files';
const PLATFORMS_FILES_SANDBOX_URL = exports.PLATFORMS_FILES_SANDBOX_URL = 'https://files.sandbox.checkout.com/files';
const TRANSFERS_SANDBOX_URL = exports.TRANSFERS_SANDBOX_URL = 'https://transfers.sandbox.checkout.com/transfers';
const TRANSFERS_LIVE_URL = exports.TRANSFERS_LIVE_URL = 'https://transfers.checkout.com/transfers';
const BALANCES_SANDBOX_URL = exports.BALANCES_SANDBOX_URL = 'https://balances.sandbox.checkout.com/balances';
const BALANCES_LIVE_URL = exports.BALANCES_LIVE_URL = 'https://balances.checkout.com/balances';
const REQUEST_ID_HEADER = exports.REQUEST_ID_HEADER = 'cko-request-id';
const API_VERSION_HEADER = exports.API_VERSION_HEADER = 'cko-version';
const DEFAULT_TIMEOUT = exports.DEFAULT_TIMEOUT = 15000;
const MBC_LIVE_SECRET_KEY_REGEX = exports.MBC_LIVE_SECRET_KEY_REGEX = /^sk_?(\w{8})-(\w{4})-(\w{4})-(\w{4})-(\w{12})$/;
const NAS_LIVE_SECRET_KEY_REGEX = exports.NAS_LIVE_SECRET_KEY_REGEX = /^sk_?[a-z2-7]{26}[a-z2-7*#$=]$/;
const NAS_SANDBOX_SECRET_KEY_REGEX = exports.NAS_SANDBOX_SECRET_KEY_REGEX = /^sk_sbox_?[a-z2-7]{26}[a-z2-7*#$=]$/;
const NAS_LIVE_PUBLIC_KEY_REGEX = exports.NAS_LIVE_PUBLIC_KEY_REGEX = /^pk_?[a-z2-7]{26}[a-z2-7*#$=]$/;
const NAS_SANDBOX_PUBLIC_KEY_REGEX = exports.NAS_SANDBOX_PUBLIC_KEY_REGEX = /^pk_sbox_?[a-z2-7]{26}[a-z2-7*#$=]$/;
const PAYMENT_TYPES = exports.PAYMENT_TYPES = {
  regular: 'Regular',
  recurring: 'Recurring',
  moto: 'MOTO',
  installment: 'Installment',
  unscheduled: 'Unscheduled'
};
const CURRENCIES = exports.CURRENCIES = {
  ALL: 'ALL',
  STN: 'STN',
  EEK: 'EEK',
  BHD: 'BHD',
  SCR: 'SCR',
  DJF: 'DJF',
  EGP: 'EGP',
  MDL: 'MDL',
  MZN: 'MZN',
  BND: 'BND',
  ZMK: 'ZMK',
  SHP: 'SHP',
  LBP: 'LBP',
  AWG: 'AWG',
  JMD: 'JMD',
  KES: 'KES',
  BYN: 'BYN',
  KHR: 'KHR',
  LAK: 'LAK',
  MVR: 'MVR',
  AOA: 'AOA',
  TJS: 'TJS',
  SVC: 'SVC',
  GNF: 'GNF',
  BRL: 'BRL',
  MOP: 'MOP',
  BOB: 'BOB',
  CDF: 'CDF',
  NAD: 'NAD',
  LYD: 'LYD',
  VUV: 'VUV',
  QAR: 'QAR',
  CLP: 'CLP',
  HRK: 'HRK',
  ISK: 'ISK',
  FKP: 'FKP',
  XCD: 'XCD',
  NOK: 'NOK',
  CUP: 'CUP',
  VND: 'VND',
  PEN: 'PEN',
  KMF: 'KMF',
  LVL: 'LVL',
  MMK: 'MMK',
  TRY: 'TRY',
  VEF: 'VEF',
  AUD: 'AUD',
  TWD: 'TWD',
  PKR: 'PKR',
  SLL: 'SLL',
  BGN: 'BGN',
  LRD: 'LRD',
  LKR: 'LKR',
  XAF: 'XAF',
  JOD: 'JOD',
  ANG: 'ANG',
  BSD: 'BSD',
  CAD: 'CAD',
  GIP: 'GIP',
  MNT: 'MNT',
  LTL: 'LTL',
  BBD: 'BBD',
  CLF: 'CLF',
  BWP: 'BWP',
  COP: 'COP',
  PHP: 'PHP',
  HUF: 'HUF',
  FJD: 'FJD',
  MWK: 'MWK',
  THB: 'THB',
  XPF: 'XPF',
  RSD: 'RSD',
  SAR: 'SAR',
  UYU: 'UYU',
  BZD: 'BZD',
  SYP: 'SYP',
  GMD: 'GMD',
  SZL: 'SZL',
  SBD: 'SBD',
  ETB: 'ETB',
  CHF: 'CHF',
  MXN: 'MXN',
  ARS: 'ARS',
  GTQ: 'GTQ',
  GHS: 'GHS',
  NIO: 'NIO',
  JPY: 'JPY',
  BDT: 'BDT',
  UZS: 'UZS',
  SOS: 'SOS',
  BTN: 'BTN',
  NZD: 'NZD',
  TZS: 'TZS',
  IQD: 'IQD',
  MGA: 'MGA',
  DZD: 'DZD',
  GYD: 'GYD',
  USD: 'USD',
  KWD: 'KWD',
  CNY: 'CNY',
  PYG: 'PYG',
  SGD: 'SGD',
  KZT: 'KZT',
  PGK: 'PGK',
  AMD: 'AMD',
  GBP: 'GBP',
  AFN: 'AFN',
  CRC: 'CRC',
  XOF: 'XOF',
  YER: 'YER',
  MRU: 'MRU',
  DKK: 'DKK',
  TOP: 'TOP',
  INR: 'INR',
  SDG: 'SDG',
  DOP: 'DOP',
  ZWL: 'ZWL',
  UGX: 'UGX',
  SEK: 'SEK',
  LSL: 'LSL',
  MYR: 'MYR',
  TMT: 'TMT',
  OMR: 'OMR',
  BMD: 'BMD',
  KRW: 'KRW',
  HKD: 'HKD',
  KGS: 'KGS',
  BAM: 'BAM',
  NGN: 'NGN',
  ILS: 'ILS',
  MUR: 'MUR',
  RON: 'RON',
  TND: 'TND',
  AED: 'AED',
  PAB: 'PAB',
  NPR: 'NPR',
  TTD: 'TTD',
  RWF: 'RWF',
  HTG: 'HTG',
  IDR: 'IDR',
  EUR: 'EUR',
  KYD: 'KYD',
  IRR: 'IRR',
  KPW: 'KPW',
  MKD: 'MKD',
  SRD: 'SRD',
  HNL: 'HNL',
  AZN: 'AZN',
  ERN: 'ERN',
  CZK: 'CZK',
  CVE: 'CVE',
  BIF: 'BIF',
  MAD: 'MAD',
  RUB: 'RUB',
  UAH: 'UAH',
  WST: 'WST',
  PLN: 'PLN',
  ZAR: 'ZAR',
  GEL: 'GEL',
  ZMW: 'ZMW'
};
